crux-mir-comp/src/Mir/Compositional/Convert.hs

```diff
@@ -246,3 +282,271 @@ readPartExprMaybe _sym W4.Unassigned = return Nothing
+                    Just x -> return x
+                    Nothing -> error $ "buildBitVector: expected " ++ show (BaseBVRepr w) ++
+                        ", but got " ++ show (W4.exprType bv')
+                return bv'
+      where
+        go :: forall w. (1 <= w) =>
+            NatRepr w ->
+            W4.SymExpr sym (BaseBVType w) ->
+            [W4.SymExpr sym (BaseBVType 1)] ->
+            IO (Some (W4.SymExpr sym))
+        go _ bv [] = return $ Some bv
+        go w bv (b : bs)
+          | LeqProof <- addPrefixIsLeq w (knownNat @1) = do
+            bv' <- W4.bvConcat sym bv b
+            go (addNat w (knownNat @1)) bv' bs
+
+-- | Common code for termToExpr and termToReg
+termToSValue :: forall sym t st fs.
+    (IsSymInterface sym, sym ~ W4.ExprBuilder t st fs, HasCallStack) =>
+    sym ->
+    SAW.SharedContext ->
+    Map SAW.VarIndex (Some (W4.Expr t)) ->
+    SAW.Term ->
+    IO (SAW.SValue sym)
+termToSValue sym sc varMap term = do
+    let convert (Some expr) = case SAW.symExprToValue (W4.exprType expr) expr of
+            Just x -> return x
+            Nothing -> error $ "termToExpr: failed to convert var  of what4 type " ++
+                show (W4.exprType expr)
+    ecMap <- mapM convert varMap
+    ref <- newIORef mempty
+    SAW.w4SolveBasic sym sc mempty ecMap ref mempty term
+
+-- | Convert a `SAW.Term` to a `W4.Pred`.  If the term doesn't have boolean
+-- type, this will raise an error.
+termToPred :: forall sym t st fs.
+    (IsSymInterface sym, sym ~ W4.ExprBuilder t st fs, HasCallStack) =>
+    sym ->
+    SAW.SharedContext ->
+    Map SAW.VarIndex (Some (W4.Expr t)) ->
+    SAW.Term ->
+    IO (W4.Pred sym)
+termToPred sym sc varMap term = do
+    Some expr <- termToExpr sym sc varMap term
+    case W4.exprType expr of
+        BaseBoolRepr -> return expr
+        btpr -> error $ "termToPred: got result of type " ++ show btpr ++ ", not BaseBoolRepr"
+
+-- | Convert a `SAW.Term` representing a type to a `W4.BaseTypeRepr`.
+termToType :: forall sym t st fs.
+    (IsSymInterface sym, sym ~ W4.ExprBuilder t st fs, HasCallStack) =>
+    sym ->
+    SAW.SharedContext ->
+    SAW.Term ->
+    IO (Some W4.BaseTypeRepr)
+termToType sym sc term = do
+    ref <- newIORef mempty
+    sv <- SAW.w4SolveBasic sym sc mempty mempty ref mempty term
+    tv <- case sv of
+        SAW.TValue tv -> return tv
+        _ -> error $ "termToType: bad SValue"
+    case tv of
+        SAW.VBoolType -> return $ Some BaseBoolRepr
+        SAW.VVecType w SAW.VBoolType -> do
+            Some w <- return $ mkNatRepr w
+            LeqProof <- case testLeq (knownNat @1) w of
+                Just x -> return x
+                Nothing -> error "termToPred: zero-width bitvector"
+            return $ Some $ BaseBVRepr w
+        _ -> error $ "termToType: bad SValue"
+
+
+exprToTerm :: forall sym t st fs tp m.
+    (IsSymInterface sym, sym ~ W4.ExprBuilder t st fs, MonadIO m, MonadFail m) =>
```

Comment:
This and a lot of the other, similar functions in this module seem like they could also be very broadly useful. It might even be worth having a What4-SAWCore integration package that included all these conversions back and forth. Centralizing that could make it easier to identify how to do the conversions as sparingly and effectively as possible.

Reply:
On second thought, the existing `saw-core-what4` package might be a good place for these sorts of functions.